### PR TITLE
Refactor to remove AstroConfig['_ctx']

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Removes `config._ctx` where state was stored and moves to a `AstroSettings` object. This object contains the config as well as the stateful pieces there were part of config._ctx.
- Updates all of the code to pass settings around instead of config.

## Testing

- Existing tests should pass.
- No changeset as this should not affect usage.

## Docs

N/A, internal refactor.

### CODE REVIEW

File: packages/astro/src/cli/index.ts [hunk 1]
```diff
@@ -161,14 +161,16 @@ async function runCommand(cmd: string, flags: yargs.Arguments) {
 	});
 	if (!astroConfig) return;
 	telemetry.record(event.eventCliSession(cmd, userConfig, flags));
+	let tsconfig = loadTSConfig(root);
+	let settings = createSettings(astroConfig, tsconfig?.config, tsconfig?.path);
```
Comment (natemoo-re) [inline packages/astro/src/cli/index.ts:165]: Nit: I can already see the day we'll have to refactor `createSettings` to accept an options bag. Maybe worth the refactor now?

File: packages/astro/src/integrations/index.ts [hunk 2]
```diff
@@ -34,21 +35,22 @@ async function withTakingALongTimeMsg<T>({
 }
 
 export async function runHookConfigSetup({
-	config: _config,
+	settings,
 	command,
 	logging,
 }: {
-	config: AstroConfig;
+	settings: AstroSettings;
```
Comment (natemoo-re) [inline packages/astro/src/integrations/index.ts:42]: This is the only area where I'm concerned about potential regressions. Is our test coverage for these hooks sufficient?

**Review by natemoo-re — COMMENTED** [review]
Clean refactor, looks great!

Didn't see a changeset, were you planning to hold this for a `minor`? That would be my suggestion given the scope.

File: packages/astro/src/integrations/index.ts (hunk 2 above)
Comment (matthewp) [reply packages/astro/src/integrations/index.ts:42]: This doesn't change the signature of hooks, this is just the function that calls the hooks.

**Comment by matthewp** [comment]
@natemoo-re No changeset since this is not a bug fix or a new feature and shouldn't effect any code. So there is no reason for a user to need to upgrade their app for this.

We can hold for minor if you want but my concerns would be that 1) this might be difficult to keep rebased since it touches so much code and 2) merging this before the minor might block other PRs that would then need to update. I thought sneaking this in a week before the minor would be a nice timing as fewer people are working on core right now.

File: packages/astro/src/cli/index.ts (hunk 1 above)
Comment (matthewp) [reply packages/astro/src/cli/index.ts]: Done here: https://github.com/withastro/astro/pull/4771/commits/b515843e4188e320d0916473289b243bb5d38aa2

**Comment by FredKSchott** [comment]
Nice! Since the point of these refactorings is the improve our code testability, can we add unit tests for the new `AstroSettings`? Are there any integration tests that are testing one-off settings, that could make more sense even now to migrate to unit tests?

**Comment by matthewp** [comment]
Yeah there are some of those. There are config validation tests which are already use tests, I'll move these into a special folder about this new module.

**Review by natemoo-re — APPROVED** [review]
Awesome, understood that this is purely an internal refactor! Merge whenever you're ready.

**Review by bluwy — APPROVED** [review]
This LGTM too! I would usually prefer to put a changeset for refactors too in case for an uncaught regression somewhere, but I don't mind if not too if we're confident in the refactors.

**Pull request merged**